src/main/java/org/creekservice/kafka/test/perf/implementations/DevHarrelImplementation.java

```diff
@@ -0,0 +1,146 @@
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        final Validator validator2019 =
+                new dev.harrel.jsonschema.ValidatorFactory()
+                        .withDialect(new Dialects.Draft2019Dialect())
+                        .withSchemaResolver(schemaResolver)
+                        .createValidator();
+        /* Validate against meta-schemas in order to parse them eagerly */
+        validator2020.validate(URI.create(SpecificationVersion.DRAFT2020_12.getId()), "{}");
+        validator2019.validate(URI.create(SpecificationVersion.DRAFT2019_09.getId()), "{}");
+
+        validator2020.registerSchema(testSchemaUri, TestSchemas.DRAFT_2020_SCHEMA);
+
+        this.validators =
+                Map.of(
+                        DRAFT_2020_12, validator2020,
+                        DRAFT_2019_09, validator2019);
+    }
+
+    @Override
+    public MetaData metadata() {
+        return METADATA;
+    }
+
+    @Override
+    public JsonValidator prepare(
+            final String schema, final SchemaSpec spec, final AdditionalSchemas additionalSchemas) {
+        DevHarrelImplementation.this.remotes = additionalSchemas.remotes();
+        final Validator validator = validators.get(spec);
+        final URI schemaUri = validator.registerSchema(schema);
+
+        return new JsonValidator() {
+            @Override
+            public void validate(final String json) {
+                final Validator.Result result = validator.validate(schemaUri, json);
+                if (!result.isValid()) {
+                    throw new RuntimeException();
+                }
+            }
+
+            @Override
+            public byte[] serialize(final TestModel model, final boolean validate) {
+                try {
+                    final String asString = mapper.writeValueAsString(model);
+                    if (validate
+                            && !validators
+                                    .get(DRAFT_2020_12)
+                                    .validate(testSchemaUri, asString)
+                                    .isValid()) {
+                        throw new RuntimeException();
+                    }
+                    return asString.getBytes(StandardCharsets.UTF_8);
+                } catch (JsonProcessingException e) {
+                    throw new RuntimeException(e);
+                }
+            }
+
+            @Override
+            public TestModel deserialize(final byte[] data) {
+                try {
+                    if (!validators
+                            .get(DRAFT_2020_12)
+                            .validate(testSchemaUri, new String(data, StandardCharsets.UTF_8))
+                            .isValid()) {
+                        throw new RuntimeException();
+                    }
+                    return mapper.readValue(data, TestModel.class);
```

Comment:
Might be quicker to avoid double string decoding:


```suggestion
                    final String json = new String(data, StandardCharsets.UTF_8);
                    if (!validators
                            .get(DRAFT_2020_12)
                            .validate(testSchemaUri, json)
                            .isValid()) {
                        throw new RuntimeException();
                    }
                    return mapper.readValue(json, TestModel.class);
```

Reply:
yep, good point